base/applications/cmdutils/attrib/attrib.c

```diff
@@ -77,217 +61,286 @@ ErrorMessage(
-                  szFullName);
-        bFound = TRUE;
-    }
-    while (FindNextFileW(hFind, &findData));
-    FindClose(hFind);
+    /* The full path buffer the function will act recursively */
+    // PWSTR FullPath;
+    size_t cchBuffer; // Buffer size
+    WCHAR FullPathBuffer[MAX_PATH + _countof("\\" ALL_FILES_PATTERN)];
 
-    return bFound;
-}
+} ENUMFILES_CTX, *PENUMFILES_CTX;
 
+unsigned int iRecurse = 0;
 
-/* Returns TRUE if anything changed, FALSE otherwise */
+/* Returns TRUE if anything is done, FALSE otherwise */
 static
 BOOL
-ChangeAttribute(
-    LPWSTR pszPath,
-    LPWSTR pszFile,
-    BOOL  bRecurse,
-    BOOL  bDirectories,
-    DWORD dwMask,
-    DWORD dwAttrib)
+EnumFilesWorker(
+    _Inout_ off_t offFilePart, // Offset to the file name inside FullPathBuffer
+    _Inout_ PENUMFILES_CTX EnumCtx)
 {
-    WIN32_FIND_DATAW findData;
+    BOOL bFound = FALSE;
+    HRESULT hRes;
     HANDLE hFind;
-    WCHAR  szFullName[MAX_PATH];
-    LPWSTR pszFileName;
-    BOOL   bFound = FALSE;
-    BOOL   bIsDir;
-    BOOL   bExactMatch;
-    DWORD  dwAttribute;
-    DWORD  Error;
-
-    /* prepare full file name buffer */
-    wcscpy(szFullName, pszPath);
-    pszFileName = szFullName + wcslen(szFullName);
-
-    /* display all subdirectories */
-    if (bRecurse)
+    PWSTR findFileName = EnumCtx->findData.cFileName;
+    PWSTR pFilePart = EnumCtx->FullPathBuffer + offFilePart;
+    size_t cchRemaining = EnumCtx->cchBuffer - offFilePart;
+
+    DEBUG_TMP PWSTR pPath = EnumCtx->FullPathBuffer;
+
+    /* Recurse over all subdirectories */
+    if (EnumCtx->Flags & ENUM_RECURSE)
     {
-        /* append *.* */
-        wcscpy(pszFileName, L"*.*");
+DEBUG_TMP
+#if 1
+        if (wcslen(pPath) > MAX_PATH-5)
+        {
+            /* NULL-terminate where we're going to append a new
+             * file part, and set it as the current directory */
+            *pFilePart = UNICODE_NULL;
+            SetCurrentDirectoryW(pPath);
+            /* and set the "full path" pointer to where we are
+             * going to set the new file part */
+            pPath = pFilePart;
+        }
+#endif
+
+        /* Append *.* */
+        hRes = StringCchCopyW(pFilePart, cchRemaining, ALL_FILES_PATTERN);
+        if (hRes != S_OK)
+        {
+            if (hRes == STRSAFE_E_INSUFFICIENT_BUFFER)
+            {
+                // TODO: If this fails, try to reallocate EnumCtx->FullPathBuffer by
+                // increasing its size by _countof(EnumCtx->findData.cFileName) + 1
+                // to satisfy this copy, as well as the one made in the loop below.
+            }
+            // else
+            ConPrintf(StdErr, L"Directory level too deep: %s\n", EnumCtx->FullPathBuffer);
+            return FALSE;
+        }
 
-        hFind = FindFirstFileW(szFullName, &findData);
+        hFind = FindFirstFileW(/*EnumCtx->FullPathBuffer*/pPath, &EnumCtx->findData);
```

Comment:
I am not sure you should re-use the `findData` while another one is still in progress?